### PR TITLE
Added Decoration, Type, and utils.

### PR DESCRIPTION
**Pull request by ehsannas**
* Added support for decorations and types.
* Introduced clang::spirv::utils
* Avoid static members in Type and Decoration based on design doc
comments. Types and Decorations are unique in the context in which
they are defined. Therefore, each context should track its own types, etc.
* getTypeXXX and getDecorationXXX functions should take a context
argument (similar to LLVM).
* Added a number of unit tests (TODO: more unit tests).

### CODE REVIEW

File: tools/clang/lib/SPIRV/Utils.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+namespace utils {
+
+/// \brief Reinterprets a given string as sequence of words.
+/// Assumes Little Endian architecture.
+std::vector<uint32_t> reinterpretStringAsUintVec(std::string s) {
+  std::vector<uint32_t> results;
+  if (!s.empty()) {
+    unsigned int num_bytes_needed = s.size() + 1;
+    unsigned int num_words_needed = (num_bytes_needed / 4) + 1;
+    std::string full_str(4 * num_words_needed, '\0');
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/Utils.cpp]: Why creating a new string instead of manipulating the original one?

File: tools/clang/lib/SPIRV/Utils.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    full_str.insert(full_str.begin(), s.begin(), s.end());
+    for (unsigned int word_index = 0; word_index < num_words_needed;
+         ++word_index) {
+      int i = 4 * word_index;
+      uint32_t word = 0;
+      word |= full_str[i + 3] << 24;
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/Utils.cpp]: Is the size of `full_str` guaranteed to be a multiple of `num_words_needed`?

File: tools/clang/lib/SPIRV/Utils.cpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  return results;
+}
+
+/// \brief Reinterprets the given vector of 32-bit words as a string.
+/// Expectes that the words represent a NULL-terminated string.
+/// Assumes Little Endian architecture.
+std::string reinterpretUintVecAsString(std::vector<uint32_t>& vec) {
+  std::string result;
+  if (!vec.empty()) {
+    result = std::string(reinterpret_cast<const char*>(&vec[0]));
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/Utils.cpp]: `vec.data()`

File: tools/clang/unittests/SPIRV/DecorationTest.cpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+TEST(ValidateType, ValidateSimpleDecorationUniqueness) {
+  using namespace clang::spirv;
+  SPIRVContext ctx;
+  const Decoration *relaxed = Decoration::getDecorationRelaxedPrecision(ctx);
+  const Decoration *relaxed_2 = Decoration::getDecorationRelaxedPrecision(ctx);
+  EXPECT_EQ(relaxed, relaxed_2);
+}
+
+TEST(ValidateType, ValidateComplexDecorationUniqueness) {
+  using namespace clang::spirv;
```
Comment (antiagainst) [inline tools/clang/unittests/SPIRV/DecorationTest.cpp]: Can just put in the beginning to avoid specifying for each test.

File: tools/clang/unittests/SPIRV/DecorationTest.cpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+      Decoration::getDecorationBuiltIn(ctx, spv::BuiltIn::Position, 0);
+  // Applied to member 1 of a structure.
+  const Decoration *mem_1_builtin_pos =
+      Decoration::getDecorationBuiltIn(ctx, spv::BuiltIn::Position, 1);
+
+  // We should get different pointers for these decorations as they are
+  // applied to different structure members.
+  EXPECT_NE(mem_0_builtin_pos, mem_1_builtin_pos);
+}
+
```
Comment (antiagainst) [inline tools/clang/unittests/SPIRV/DecorationTest.cpp]: Tests for multiple different decorations?

File: tools/clang/unittests/SPIRV/SPIRVContextTest.cpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+#include "gmock/gmock.h"
+#include "clang/SPIRV/Decoration.h"
+#include "clang/SPIRV/SPIRVContext.h"
+#include "clang/SPIRV/Type.h"
+#include "gtest/gtest.h"
+
+namespace {
+
+TEST(ValidateSPIRVContext, ValidateGetNextId) {
```
Comment (antiagainst) [inline tools/clang/unittests/SPIRV/SPIRVContextTest.cpp]: Nice! Thanks! :)

File: tools/clang/unittests/SPIRV/UtilsTest.cpp [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+//===----------------------------------------------------------------------===//
+
+#include "clang/SPIRV/Utils.h"
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+
+namespace {
+
+using ::testing::ElementsAre;
+
```
Comment (antiagainst) [inline tools/clang/unittests/SPIRV/UtilsTest.cpp]: Need tests for empty string, string with one characters, two characters, three characters, and four characters. These are important corner cases.

File: tools/clang/unittests/SPIRV/SPIRVContextTest.cpp [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+TEST(ValidateSPIRVContext, ValidateTakeNextId) {
+  clang::spirv::SPIRVContext context;
+  EXPECT_EQ(context.TakeNextId(), 1u);
+  EXPECT_EQ(context.TakeNextId(), 2u);
+  EXPECT_EQ(context.GetNextId(), 3u);
+}
+
+TEST(ValidateSPIRVContext, ValidateUniqueIdForUniqueNonAggregateType) {
+  using namespace clang::spirv;
```
Comment (antiagainst) [inline tools/clang/unittests/SPIRV/SPIRVContextTest.cpp]: Just need to put in the beginning to avoid specifying for each test.

File: tools/clang/unittests/SPIRV/SPIRVContextTest.cpp [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  // We also assign RelaxedPrecision decoration to the struct as a whole.
+  // We also assign BufferBlock decoration to the struct as a whole.
+  // We also assign Offset decoration to each member of the struct.
+  // We also assign a BuiltIn decoration to the first member of the struct.
+  const Type *intt = Type::getTypeInt32(ctx);
+  const Type *boolt = Type::getTypeBool(ctx);
+  const uint32_t intt_id = ctx.GetResultIdForType(intt, NextIdFunc);
+  const uint32_t boolt_id = ctx.GetResultIdForType(boolt, NextIdFunc);
+  const Decoration *relaxed = Decoration::getDecorationRelaxedPrecision(ctx);
+  const Decoration *bufferblock = Decoration::getDecorationBufferBlock(ctx);
```
Comment (antiagainst) [inline tools/clang/unittests/SPIRV/SPIRVContextTest.cpp]: Can just use `const auto ... = ...` here since the rhs already clearly shows that's the returned type.

File: tools/clang/unittests/SPIRV/DecorationTest.cpp [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+//===----------------------------------------------------------------------===//
+
+#include "gmock/gmock.h"
+#include "clang/SPIRV/Decoration.h"
+#include "clang/SPIRV/SPIRVContext.h"
+#include "gtest/gtest.h"
+
+namespace {
+
+TEST(ValidateType, ValidateSimpleDecorationUniqueness) {
```
Comment (antiagainst) [inline tools/clang/unittests/SPIRV/DecorationTest.cpp]: I typically don't really like to have `Validate` or `Check` or similar words in the test name since test are surely validating/checking something. Having them is not providing more information but takes the space. Instead, test name should be clear what's the condition and what's the expected outcome (from style guide). What about something like ```Test(Type, SameDecorationWoParameterShouldHaveSameAddress)```?

Also applies to the following test cases.

File: tools/clang/unittests/SPIRV/SPIRVContextTest.cpp [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+  const Type *struct_2 = Type::getType(
+      ctx, spv::Op::OpTypeStruct, {intt_id, boolt_id},
+      {relaxed, bufferblock, mem_0_offset, mem_1_offset, mem_0_position});
+
+  const uint32_t struct_1_id = ctx.GetResultIdForType(struct_1, NextIdFunc);
+  const uint32_t struct_2_id = ctx.GetResultIdForType(struct_2, NextIdFunc);
+
+  // We should be able to retrieve the same ID for the same Type.
+  EXPECT_EQ(struct_1_id, struct_2_id);
+}
```
Comment (antiagainst) [inline tools/clang/unittests/SPIRV/SPIRVContextTest.cpp]: Need test case for
* vectors of different types/number-of-elements
* structs with different struct decorations
* structs with different member decorations
* etc.

File: tools/clang/lib/SPIRV/Utils.cpp [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+#include "clang/SPIRV/Utils.h"
+#include "llvm/llvm_assert/assert.h"
+
+namespace clang {
+namespace spirv {
+namespace utils {
+
+/// \brief Reinterprets a given string as sequence of words.
+/// Assumes Little Endian architecture.
+std::vector<uint32_t> reinterpretStringAsUintVec(std::string s) {
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/Utils.cpp]: FYI: I have implemented this function in #15 since I need it for additional works. It uses similar logic as https://github.com/KhronosGroup/SPIRV-Tools/blob/f18e1f293b7cf8f24a42587b48f35bc49912a5dd/source/text_handler.cpp#L273. Feel free to move the implementation from #15 to this file.

File: tools/clang/include/clang/SPIRV/SPIRVContext.h [hunk 13]
```diff
@@ -30,8 +47,35 @@ class SPIRVContext {
...
+  // All the unique types defined in the current context.
+  std::unordered_set<Type, TypeHash> ExistingTypes;
+
+  // All the unique Decorations defined in the current context.
+  std::unordered_set<Decoration, DecorationHash> ExistingDecorations;
+
 private:
   uint32_t NextId;
+
+  // Maps a ResultID to its SPIR-V instruction.
```
Comment (ehsannas) [inline tools/clang/include/clang/SPIRV/SPIRVContext.h]: Yes, I agree 👍

File: tools/clang/include/clang/SPIRV/Type.h [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+#ifndef LLVM_CLANG_SPIRV_TYPE_H
+#define LLVM_CLANG_SPIRV_TYPE_H
+
+#include <set>
+#include <unordered_set>
```
Comment (ehsannas) [inline tools/clang/include/clang/SPIRV/Type.h]: we actually need \<set\> because we use an _ordered set_ of Decorations in a Type.
Inserting into an ordered set is less efficient than inserting into an unordered_set. (log(n) vs constant). But, having ordered sets makes comparison of Type objects much more efficient.
Note that insertion occurs only _once_ when the type is created, however, comparison could occur many many times.

File: tools/clang/include/clang/SPIRV/SPIRVContext.h [hunk 15]
```diff
@@ -30,8 +47,35 @@ class SPIRVContext {
   inline uint32_t GetNextId() const;
   inline uint32_t TakeNextId();
 
+  /// \brief Returns the ResultID that defines the given Type. If the type
+  /// has not been defined, it will define and store its instruction.
+  /// The 'next_id_func' will be called to get the ResultID for the type.
+  uint32_t GetResultIdForType(const Type *t,
```
Comment (ehsannas) [inline tools/clang/include/clang/SPIRV/SPIRVContext.h]: good suggestion

File: tools/clang/include/clang/SPIRV/Type.h [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+namespace clang {
+namespace spirv {
+
+class SPIRVContext;
+
+class Type {
+public:
+  spv::Op getOpcode() const { return opcode; }
+  const std::vector<uint32_t> &getArgs() const { return args; }
+  bool isBooleanType(const Type *t);
```
Comment (ehsannas) [inline tools/clang/include/clang/SPIRV/Type.h]: yeah... i don't know how this happened! looks like an old version of the code got mixed in :)

File: tools/clang/include/clang/SPIRV/Utils.h [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+#include <string>
+#include <vector>
+
+namespace clang {
+namespace spirv {
+namespace utils {
+
+/// \brief Reinterprets a given string as sequence of words.
+/// Assumes Little Endian architecture.
```
Comment (ehsannas) [inline tools/clang/include/clang/SPIRV/Utils.h]: aah! i see 👍

File: tools/clang/lib/SPIRV/Decoration.cpp [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+#include "llvm/llvm_assert/assert.h"
+
+namespace clang {
+namespace spirv {
+
+const Decoration *Decoration::getUniqueDecoration(SPIRVContext &context,
+                                                  Decoration &d) {
+  // Insert function will only insert if it doesn't already exist in the set.
+  context.ExistingDecorations.insert(d);
+  return &(*context.ExistingDecorations.find(d));
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/Decoration.cpp]: very cool! +1

File: tools/clang/unittests/SPIRV/DecorationTest.cpp (hunk 5 above)
Comment (ehsannas) [inline tools/clang/unittests/SPIRV/DecorationTest.cpp]: I will be adding many more tests in the next PR. This CL is already quite large.

File: tools/clang/unittests/SPIRV/SPIRVContextTest.cpp (hunk 11 above)
Comment (ehsannas) [inline tools/clang/unittests/SPIRV/SPIRVContextTest.cpp]: Will do in the next CL. Opened a new issue for it.

File: tools/clang/include/clang/SPIRV/SPIRVContext.h [hunk 19]
```diff
@@ -27,16 +44,47 @@ class SPIRVContext {
...
+  uint32_t getResultIdForType(const Type *);
+
+  /// \brief Returns the instruction that defined the given Type.
+  /// If no <result-id> has been associated with this Type (not defined yet),
+  /// the instruction will be created, registered, and returned.
+  const std::vector<uint32_t> &getInstrForType(const Type *);
+
+  /// \brief Registers the existence of the given type in the current context,
+  /// and returns the unique Type pointer.
+  const Type *registerType(Type &);
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/SPIRVContext.h]: Should use `const &` if we are not gonna change the passed in type.

File: tools/clang/include/clang/SPIRV/SPIRVContext.h [hunk 20]
```diff
@@ -27,16 +44,47 @@ class SPIRVContext {
...
+
+  /// \brief Registers the existence of the given decoration in the current
+  /// context, and returns the unique Decoration pointer.
+  const Decoration *registerDecoration(Decoration &);
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/SPIRVContext.h]: Should use `const &` if we are not gonna change the passed in decoration.

File: tools/clang/include/clang/SPIRV/Type.h (hunk 14 above)
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/Type.h]: Makes sense.

File: tools/clang/include/clang/SPIRV/Type.h [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+///
+/// This class defines a unique SPIR-V Type.
+/// A SPIR-V Type includes its <opcode> defined by the SPIR-V Spec.
+/// It also incldues any arguments (32-bit words) needed to define the
+/// type. It also includes a set of decorations that are applied to that type.
+///
+/// The class includes static getXXX(...) functions for getting pointers of any
+/// needed type. A unique type has a unique pointer (e.g. calling
+/// 'getBoolean' function will always return the same pointer for the given
+/// context)
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/Type.h]: dot to end this sentence.

File: tools/clang/include/clang/SPIRV/Type.h [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+class Type {
+public:
+  spv::Op getOpcode() const { return opcode; }
+  const std::vector<uint32_t> &getArgs() const { return args; }
+  bool isBooleanType() const;
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/Type.h]: Empty line to separate `isXXXType` methods with the above.

File: tools/clang/include/clang/SPIRV/Type.h [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+  bool isScalarType() const;
+  bool isVectorType() const;
+  bool isMatrixType() const;
+  bool isArrayType() const;
+  bool isStructureType() const;
+  bool isAggregateType() const;
+  bool isCompositeType() const;
+  bool isImageType() const;
+
+  static const Type *getType(SPIRVContext &ctx, spv::Op op,
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/Type.h]: Do we still need a public `getType` method here?

File: tools/clang/include/clang/SPIRV/Decoration.h [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+  Decoration(spv::Decoration dec_id, llvm::SmallVector<uint32_t, 2> arg = {},
+    llvm::Optional<uint32_t> idx = llvm::None)
+      : id(dec_id), args(arg), memberIndex(idx) {}
+
+  /// \brief Sets the index of the structure member to which the decoration
+  /// applies.
+  void setMemberIndex(llvm::Optional<uint32_t> idx) { memberIndex = idx; }
+
+  /// \brief Returns the unique decoration pointer within the given context.
+  static const Decoration *getUniqueDecoration(SPIRVContext &ctx,
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/Decoration.h]: Should use `const &` if we are not gonna change the Decoration passed in.

File: tools/clang/include/clang/SPIRV/Type.h [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+           decorations == other.decorations;
+  }
+
+private:
+  /// \brief Private constructor.
+  Type(spv::Op op, std::vector<uint32_t> arg = {},
+       std::set<const Decoration *> dec = {});
+
+  /// \brief Returns the unique Type pointer within the given context.
+  static const Type *getUniqueType(SPIRVContext &, Type &);
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/Type.h]: Should use `const &` if we are not gonna change the Type passed in.

File: tools/clang/lib/SPIRV/SPIRVContext.cpp [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+
+const std::vector<uint32_t> &
+SPIRVContext::getInstrForType(const Type *t) {
+  uint32_t result_id = getResultIdForType(t);
+  return idToInstructionMap[result_id];
+}
+
+const Type* SPIRVContext::registerType(Type& t) {
+  // Insert function will only insert if it doesn't already exist in the set.
+  std::unordered_set<Type, TypeHash>::iterator it;
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVContext.cpp]: Can actually define a private `using TypeSet = std::unordered_set<Type, TypeHash>` in `SPIRVContext`. Will make this simpler to read. Also for Decoration.

File: tools/clang/unittests/SPIRV/UtilsTest.cpp [hunk 27]
```diff
@@ -0,0 +1,91 @@
...
+TEST(Utils, EncodeEmptyString) {
+  std::string str = "";
+  std::vector<uint32_t> words =
+      clang::spirv::utils::encodeSPIRVString(str);
```
Comment (antiagainst) [inline tools/clang/unittests/SPIRV/UtilsTest.cpp]: Can use namespace clang::spirv to shorten these. These are tests; using namespaces should be fine.

File: tools/clang/include/clang/SPIRV/SPIRVContext.h [hunk 28]
```diff
@@ -27,16 +44,47 @@ class SPIRVContext {
...
+  inline uint32_t takeNextId();
+
+  /// \brief Returns the <result-id> that defines the given Type. If the type
+  /// has not been defined, it will define and store its instruction.
+  uint32_t getResultIdForType(const Type *);
+
+  /// \brief Returns the instruction that defined the given Type.
+  /// If no <result-id> has been associated with this Type (not defined yet),
+  /// the instruction will be created, registered, and returned.
+  const std::vector<uint32_t> &getInstrForType(const Type *);
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/SPIRVContext.h]: I still don't quite think we should have this method in SPIRVContext. Constructing concrete SPIR-V words should be the responsibility of SPIR-V module and inst builder. SPIRVContext should be only in charge of id assignment and resource allocation etc..

IMHO we just need to keep the functionality of assigning `<result-id>`s for the types in SPIRVContext, via `registerType`. `registerType` should automatically reserve a `<result-id>` for the given type if new. No need to actually constructing the SPIR-V words for the type.

How do you think?

**Comment by antiagainst** [comment]
Almost ready to go! Just a few nits. :)

File: tools/clang/include/clang/SPIRV/Decoration.h [hunk 29]
```diff
@@ -9,145 +9,152 @@
...
+  static const Decoration *getOverrideCoverageNV(SPIRVContext &ctx);
+  static const Decoration *getPassthroughNV(SPIRVContext &ctx);
+  static const Decoration *getViewportRelativeNV(SPIRVContext &ctx);
+  static const Decoration *getSecondaryViewportRelativeNV(SPIRVContext &ctx,
+                                                          uint32_t offset);
 
   bool operator==(const Decoration &other) const {
     return id == other.id && args == other.args &&
-           memberIndex == other.memberIndex;
+           memberIndex.hasValue() == other.memberIndex.hasValue() &&
```
Comment (ehsannas) [inline tools/clang/include/clang/SPIRV/Decoration.h]: according to Optional.h, you can't compare two Optional objects.
```cpp
/// \brief Poison comparison between two \c Optional objects. Clients needs to
/// explicitly compare the underlying values and account for empty \c Optional
/// objects.
///
/// This routine will never be defined. It returns \c void to help diagnose
/// errors at compile time.
template<typename T, typename U>
void operator==(const Optional<T> &X, const Optional<U> &Y);
```

File: tools/clang/include/clang/SPIRV/Decoration.h [hunk 30]
```diff
@@ -9,145 +9,152 @@
...
-                                                  uint32_t alignment);
-  static const Decoration *getDecorationMaxByteOffset(SPIRVContext &ctx,
-                                                      uint32_t max_byte_offset);
-  static const Decoration *getDecorationOverrideCoverageNV(SPIRVContext &ctx);
-  static const Decoration *getDecorationPassthroughNV(SPIRVContext &ctx);
-  static const Decoration *getDecorationViewportRelativeNV(SPIRVContext &ctx);
-  static const Decoration *
-  getDecorationSecondaryViewportRelativeNV(SPIRVContext &ctx, uint32_t offset);
+  spv::Decoration getValue() const { return id; }
+  const llvm::SmallVectorImpl<uint32_t> &getArgs() const { return args; }
```
Comment (ehsannas) [inline tools/clang/include/clang/SPIRV/Decoration.h]: Both work actually. I'll update it to make it consistent.

File: tools/clang/include/clang/SPIRV/Type.h (hunk 23 above)
Comment (ehsannas) [inline tools/clang/include/clang/SPIRV/Type.h]: we still do since API users may want to create arbitrary Type objects with arbitrary decorations that they want. However, most often users will be using getInt, getBool, getFloat, etc.

File: tools/clang/include/clang/SPIRV/SPIRVContext.h (hunk 28 above)
Comment (ehsannas) [inline tools/clang/include/clang/SPIRV/SPIRVContext.h]: sounds reasonable 👍

File: tools/clang/include/clang/SPIRV/Decoration.h (hunk 29 above)
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/Decoration.h]: Oh, I see. Should read it before leaving this comment. :)

File: tools/clang/include/clang/SPIRV/Type.h (hunk 23 above)
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/Type.h]: OK.

Comment (ehsannas) [inline tools/clang/include/clang/SPIRV/Type.h]: I have addressed this comment in the pull #27 
I have removed the `getType` function to prevent users from constructing invalid types. In order to compensate for that, I updated all Type::get...() functions so they can accept decorations.
